R-packages/evalcast/R/download_signals.R

```diff
@@ -1,30 +1,104 @@
+#' e.g. c("ak", "ca") for states.
+#' @param as_of a date in the format "YYYY-MM-DD". Pulls the latest issue of the data available in covidcast on this date.
+#' @param offline_signal_dir the directory that stores the cached data for each (signal, forecast day) pair. If this is
+#' null, no caching is done and the data is downloaded from covidcast. The data is stored in a csv file with the format
+#' 'offline_signal_dir/data-source_signal_geo-type_as-of.csv`. Warning: no intelligent fetching is done, so if you request
+#' the same signal, geo type, and as of, but vary the other parameters (non-empty symmetric difference geo_value subsets or
+#' non-empty symmetric difference start_day and end_day arguments), then the cache will return the result of the first call
+#' for later calls. To avoid this and use caching, make sure to make a call that fetches all the data you anticipate to use
+#' first and then request subsets. You can use populate_cache() below to fetch all the needed data.
+#' @param ... other arguments to be passed to `covidcast::covidcast_signal()`.
 #' @return `covidcast_signal` data frame.
 #'
-download_signal <- function(...) {
-  args <- list(...)
-  if (is.null(args$start_day)) {
-    msg <- stringr::str_glue(
-      "Downloading {signal} from covidcast through {end}.",
-      start = args$start_day,
-      end = args$end_day,
-      signal = args$signal,
-      .sep = " "
-    )
+#' @importFrom assertthat assert_that
+#' @importFrom stringr str_glue
+#' @importFrom rlang inform warn
+#' @importFrom fs dir_create
+#' @importFrom magrittr %>%
+#' @importFrom plyr empty
+download_signal <- function(data_source, signal, start_day = NULL, end_day = NULL, geo_type = "county", geo_values = "*", as_of = NULL, offline_signal_dir = NULL, ...) {
+  if (is.null(geo_type)) geo_type <- "county"
+  if (is.null(geo_values)) geo_values <- "*"
+  if (is.null(as_of)) as_of <- Sys.Date()
+  if (is.null(end_day)) end_day <- max(as_of)
+  if (!is.null(start_day)) assert_that(as.Date(start_day) < as.Date(end_day))
+
+  if (is.null(offline_signal_dir)) {
+    msg <- str_glue("Downloading {signal} from covidcast from ", signal = signal, .sep = " ")
+
+    if (is.null(start_day)) msg <- str_glue("{msg} the earliest date possible to", msg = msg)
+    else msg <- str_glue("{msg} {start_day} to", msg = msg, start_day = start_day)
+
+    if (is.null(end_day)) msg <- str_glue("{msg} the latest date possible.", msg = msg)
+    else msg <- str_glue("{msg} {end_day}", msg = msg, end_day = end_day)
+
+    message(msg)
+    df <- suppressMessages(covidcast_signal_wrapper(data_source = data_source, signal = signal, start_day = start_day, end_day = end_day, geo_type = geo_type, geo_values = geo_values, as_of = as_of, ...))
+  } else {
+    signal_fpath <- file.path(offline_signal_dir, sprintf("%s_%s_%s_%s.RDS", data_source, signal, geo_type, as_of))
```

Comment:
Does `time_type` need to be here too?  Or is it limited to be `"day"` somewhere else?

Reply:
Looking at `get_predictions` it looks like there are no protections for incidence period being day or week. It's all treated the same way. However, `get_predictions_single_date` has a line that converts its `forecast_date` argument using `lubridate::ymd(forecast_date)`, which seems like it has a hidden assumption that the provided dates are not epiweeks. That codepath seems _very_ shaky and we're probably just lucky that no one ever used `get_predictions` with epiweeks. 

To address this, we can probably include time_type, but treat caching in the same way.

Reply:
Right, IIRC `get_predictions` weekly predictions are based on `time_type="day"` and `incidence_period="week"`, and probably would blow up with `time_type="week"`.  So either the recording `time_type` or hard-coding `time_type="day"` seems sufficient.

Reply:
It looks like `covidcast::covidcast_signal` takes only YYYYMMDD format dates for `start_day` and `end_day`, so I'm going to keep those same requirements here. However in the comparison below, I'm going to convert `start_day` and `end_day` to epiweek format so we can do the time comparison with the response (I'm assuming the response time_value is in epiweek format, will need to double check).

Reply:
I'm guessing that evalcast will break elsewhere if you get `time_type="week"` downloads and caching working, so `get_forecasts` will need to check for and complain if `time_type="week"` to break "nicely" with a complaint.  If it seems like substantial work to get the weekly downloading or caching working, might as well just complain there instead if someone tries downloading `time_type="week"` data (it wouldn't be `get_predictions` requesting it --- earlier I forgot how incidence_period="epiweek" worked).

Reply:
FWIW, I just tested this and it turns out that `time_type='week'` signals still return regular dates.

```r
r$> d <- covidcast::covidcast_signal("nchs-mortality", "deaths_allcause_incidence_num", geo_type = "state", start_day = "2021-01-01", time_type = "week")
r$> d %>% tibble
# A tibble: 3,744 × 12
   data_source    signal                        geo_value time_value issue        lag missing_value missing_stderr missing_sample_size value stderr sample_size
   <chr>          <chr>                         <chr>     <date>     <date>     <int>         <int>          <int>               <int> <dbl> <lgl>  <lgl>      
 1 nchs-mortality deaths_allcause_incidence_num ak        2020-12-27 2021-10-31    44             0              1                   1   105 NA     NA         
 2 nchs-mortality deaths_allcause_incidence_num al        2020-12-27 2021-11-21    47             0              1                   1  1804 NA     NA         
 3 nchs-mortality deaths_allcause_incidence_num ar        2020-12-27 2021-11-28    48             0              1                   1   967 NA     NA
```

But sure, if you think that evalcast was just never really intended to be used with `time_type='week'` and it hasn't been, then we can just make that explicit by having it fail with an explanation.

Reply:
I think it actually silently ignores `time_type="week"` I believe (any `time_type` arg doesn't make its way to `download_signal`).  For weekly stuff the `evalcast` user was expected to use `incidence_period="epiweek"`.